### ISSUE TITLE
AWS delete action for unused iam profiles

### ISSUE THREAD

**Issue by nitrocode** [cloud-custodian/cloud-custodian#3637]
We can already filter on unused iam profiles but cannot delete yet. This ticket is to add delete functionality which may require a 2 step process, dissociate role from profile, and then delete or just delete may automatically dissociate the role.

**Comment by coingraham** [comment]
I don't think you need to disassociate the iam role in order to delete, but deleting an instance profile that's attached to any ec2 instance will break things.  @kapilt, what would you think about creating an instance profile delete action but require the "unused" filter?

**Comment by kapilt** [comment]
i'm a little cautious about being overly prescriptive, ie. say you were doing it as an event on create, then unused just adds a lot of over head to fetch extant usages of unrelated profiles. it is worth noting in the docs about effects on extant references/usages.

**Comment by coingraham** [comment]
Just concerned that deleting an instance profile could causes irrecoverable damage to instances if they were attached.  What about code to check before deletion?

**Comment by nitrocode** [comment]
> I don't think you need to disassociate the iam role in order to delete, but deleting an instance profile that's attached to any ec2 instance will break things. @kapilt, what would you think about creating an instance profile delete action but require the "unused" filter?

I ran into an issue today where I tried to remove an unused iam profile that contained a role. I had to first remove the role it was associated with and then remove the profile itself.
